apps/hash-api/src/graph/migrate-ontology-types/util.ts

```diff
@@ -0,0 +1,337 @@
+      context,
+      authentication,
+      {
+        ownedById: accountGroupId as OwnedById,
+        schema: propertyTypeSchema,
+        webShortname: systemTypeWebShortname,
+      },
+    ).catch((createError) => {
+      // logger.warn(`Failed to create property type: ${propertyTypeId}`);
+      throw createError;
+    });
+
+    // We don't want anyone but the systemAccount being able to modify system types
+    await modifyPropertyTypeAuthorizationRelationships(
+      context,
+      authentication,
+      [
+        {
+          operation: "delete",
+          relationship: {
+            relation: "owner",
+            resource: {
+              kind: "propertyType",
+              resourceId: createdPropertyType.schema.$id,
+            },
+            subject: {
+              kind: "accountGroup",
+              subjectId: accountGroupId,
+            },
+          },
+        },
+        {
+          operation: "create",
+          relationship: {
+            relation: "owner",
+            resource: {
+              kind: "propertyType",
+              resourceId: createdPropertyType.schema.$id,
+            },
+            subject: {
+              kind: "account",
+              subjectId: systemAccountId,
+            },
+          },
+        },
+      ],
+    );
+
+    return createdPropertyType;
+  }
+};
+
+export const createSystemEntityTypeIfNotExists: ImpureGraphFunction<
+  {
+    entityTypeDefinition: {
+      allOf?: VersionedUrl[];
+      title: string;
+      description?: string;
+      properties?: {
+        propertyType: PropertyTypeWithMetadata | VersionedUrl;
+        required?: boolean;
+        array?: { minItems?: number; maxItems?: number } | boolean;
+      }[];
+      outgoingLinks?: {
+        linkEntityType: EntityTypeWithMetadata | VersionedUrl;
+        destinationEntityTypes?: [
+          LinkDestinationConstraint,
+          ...LinkDestinationConstraint[],
+        ];
+        minItems?: number;
+        maxItems?: number;
+        ordered?: boolean;
+      }[];
+    };
+    migrationState: MigrationState;
+  },
+  Promise<EntityTypeWithMetadata>
+> = async (
+  context,
+  authentication,
+  { entityTypeDefinition, migrationState },
+) => {
+  const { title } = entityTypeDefinition;
+  const baseUrl = generateSystemTypeBaseUrl({ kind: "entity-type", title });
+
+  const versionNumber = 1;
+
+  const entityTypeId = versionedUrlFromComponents(baseUrl, versionNumber);
+
+  migrationState.entityTypeVersions[baseUrl] = versionNumber;
+
+  const existingEntityType = await getEntityTypeById(context, authentication, {
+    entityTypeId,
+  }).catch((error: Error) => {
+    if (error instanceof NotFoundError) {
+      return null;
+    }
+    throw error;
+  });
+
+  if (existingEntityType) {
+    return existingEntityType;
+  }
+
+  const entityTypeSchema = generateSystemEntityTypeSchema({
+    ...entityTypeDefinition,
+    entityTypeId,
+  });
+
+  // The type was missing, try and create it
```

Comment:
@CiaranMn there's a bunch of code duplication in this file with the `src/graph/util.ts` file, which I'm hoping we can get rid of when we no longer need the initializer logic for the linear system types. If you think it's worth it I can try and share the logic between the `createSystemEntityTypeIfNotExists` and `entityTypeInitializer`, as well as the related functions.

Reply:
Let's leave it like this for now and remove it when we update Linear